src/main/java/ch/njol/skript/expressions/ExprSignText.java

```diff
@@ -18,159 +18,298 @@
+		} else if (!lines) {
+			line = this.line.getOptionalSingle(event).orElse(1).intValue() - 1;
+			if (line < 0 || line > 3)
+				return new String[0];
 		}
-		final Block b = block.getSingle(e);
-		if (b == null)
-			return new String[0];
-		if (!sign.isOfType(b))
-			return new String[0];
-		return new String[] {((Sign) b.getState()).getLine(line)};
-	}
-	
-	@Override
-	public String toString(final @Nullable Event e, final boolean debug) {
-		return "line " + line.toString(e, debug) + " of " + block.toString(e, debug);
+		if (getTime() >= 0 && event instanceof SignChangeEvent && !Delay.isDelayed(event)) {
+			if (lines)
+				return ((SignChangeEvent) event).getLines();
+			return new String[] {((SignChangeEvent) event).getLine(line)};
+		}
+		int finalLine = line;
+		return blocks.stream(event)
+				.map(Block::getState)
+				.filter(Sign.class::isInstance)
+				.map(Sign.class::cast)
+				.flatMap(sign -> {
+					if (RUNNING_1_20) {
+						if (lines)
+							return Arrays.stream(sign.getSide(side).getLines());
+						return Stream.of(sign.getSide(side).getLine(finalLine));
+					}
+					if (lines)
+						return Arrays.stream(sign.getLines());
+					return Stream.of(sign.getLine(finalLine));
+				})
+				.toArray(String[]::new);
 	}
-	
-	// TODO allow add, remove, and remove all (see ExprLore)
+
 	@Override
 	@Nullable
-	public Class<?>[] acceptChange(final ChangeMode mode) {
-		if (mode == ChangeMode.DELETE || mode == ChangeMode.SET)
-			return new Class[] {String.class};
-		return null;
+	public Class<?>[] acceptChange(ChangeMode mode) {
+		boolean acceptsMany = line == null;
+		switch (mode) {
+			case REMOVE:
+			case REMOVE_ALL:
+			case DELETE:
+				acceptsMany = false;
+			case SET:
+			case ADD:
+				return CollectionUtils.array(acceptsMany ? String[].class : String.class);
+			case RESET:
+			default:
+				return null;
+		}
 	}
-	
-	static boolean hasUpdateBooleanBoolean = true;
-	
-	@SuppressWarnings("incomplete-switch")
+
 	@Override
-	public void change(final Event e, final @Nullable Object[] delta, final ChangeMode mode) throws UnsupportedOperationException {
-		final Number l = line.getSingle(e);
-		if (l == null)
-			return;
-		final int line = l.intValue() - 1;
-		if (line < 0 || line > 3)
-			return;
-		final Block b = block.getSingle(e);
-		if (b == null)
+	public void change(Event event, @Nullable Object[] delta, ChangeMode mode) {
+		int line = 0;
+		if (this.line == null && !lines) {
 			return;
-		if (getTime() >= 0 && e instanceof SignChangeEvent && b.equals(((SignChangeEvent) e).getBlock()) && !Delay.isDelayed(e)) {
-			switch (mode) {
-				case DELETE:
-					((SignChangeEvent) e).setLine(line, "");
-					break;
-				case SET:
-					assert delta != null;
-					((SignChangeEvent) e).setLine(line, (String) delta[0]);
-					break;
-			}
-		} else {
-			if (!sign.isOfType(b))
+		} else if (!lines) {
+			line = this.line.getOptionalSingle(event).orElse(1).intValue() - 1;
```

Comment:
Same here, this shouldn't fallback to a default value to not give unexpected results